ir-test.cxx

```diff
@@ -231,25 +233,79 @@ namespace ir {
 	}
 }
 
+static void print_help(void)
+{
+	std::cout << "Run IR uint tests\n";
+	std::cout << "Usage:\n  ";
+	std::cout << ::this_exe
+	          << " [--show-diff] [--no-color] [test folders or files...]\n";
+	std::cout << "  Run all tests if no test folders/files are specified\n";
+}
 int main(int argc, char **argv) {
+	// save this executable path to infer build_dir and test_dir
+	::this_exe = argv[0];
+	::this_exe_path = std::filesystem::path{::this_exe}.parent_path();
+
+	// store test folders/files specified by user
+	std::set<std::string> user_files;
+	bool opt_error = false;
+
 	for (int i = 1; i < argc; i++) {
 		// XXX use some cleaner arg parsing solution
 		if (!std::string(argv[i]).compare("--show-diff")) {
 			::show_diff = true;
+			continue;
 		} else if (!std::string(argv[i]).compare("--no-color")) {
 			::colorize = false;
+			continue;
+		} else if (!std::string(argv[i]).compare("--help")) {
+			print_help();
+			return 0;
+		} else {
+			// user specified test folders/files
+			std::string file = argv[i];
+			if (std::filesystem::exists(file)
+			    && (std::filesystem::is_directory(file)
+		            || std::filesystem::is_regular_file(file))) {
+				user_files.insert(argv[i]);
+			} else {
+				opt_error = true;
+				std::cerr << ir::colorize("ERROR", ir::RED)
+				     << ": [" << file << "] NOT exist or a regular file\n";
+			}
 		}
+	} /* for: arguments iteration */
+
+	if (opt_error) {
+		return 1;
 	}
+
 	ir::init_console();
 
 	::build_dir = ir::get_dir_from_env("BUILD_DIR");
 	::src_dir = ir::get_dir_from_env("SRC_DIR");
-	::test_dir = ::src_dir + PATH_SEP + "tests";
 	::ir_exe = ::build_dir + PATH_SEP + "ir" + EXE_SUF;
 	::ir_target = ir::trim(ir::exec(::ir_exe + " --target"));
+
+	// get test files, either specified by user or all tests by default
 	std::vector<std::string> irt_files;
+	if (user_files.empty()) {
+		// get all tests
+		::test_dir = ::src_dir + PATH_SEP + "tests";
+		ir::find_tests_in_dir(::test_dir, irt_files);
+	} else {
+		// get user specified tests
+		for (auto f : user_files) {
```

Comment:
Could this loop be packed into `find_tests_in_dir`? Then it would accept a `set` as the first argument instead of a `string`. The `set` would consist either on the user supplied list of filenames, or on the `test_dir` if none given. Or maybe just use another overloaded version of `find_tests_in_dir`, that would call itself with the directory/file path, depending on what the current item in `set` is? This would lay a base for a further improvement in the future to add some kind of `glob`, too.

Thanks

Reply:
good idea. I will refactor this code part.

Reply:
packed the loop into `find_tests_in_dir` and now this function accepts a `set` as its first argument.

Reply:
LGTM!